Added article "Using frames in WebView2 apps" (#2251)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/how-to/hostobject.md

ms.prod: microsoft-edge
ms.technology: webview
ms.date: 11/16/2022
---
# Call native-side code from web-side code

WebView2 enables applications to bridge the gap between the web and native sides of an application by enabling an object to be passed to the web.  You expose selected native-side APIs to your webpage JavaScript through an intermediary native host object that's defined in the native code.  The native-side APIs are projected into JavaScript by using the WebView2 `AddHostObjectToScript` API.

<!--
##### [.NET](#tab/dotnet)

* [CoreWebView2.AddHostObjectToScript Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.addhostobjecttoscript)
-->

##### [Win32](#tab/win32)

* [ICoreWebView2::AddHostObjectToScript method](/microsoft-edge/webview2/reference/win32/icorewebview2#addhostobjecttoscript)

##### [.NET in a frame](#tab/dotnetframe)

* [CoreWebView2Frame.AddHostObjectToScript](/dotnet/api/microsoft.web.webview2.core.corewebview2frame.addhostobjecttoscript)

##### [Win32 in a frame](#tab/win32frame)

* [ICoreWebView2Frame::AddHostObjectToScriptWithOrigins method](/microsoft-edge/webview2/reference/win32/icorewebview2frame#addhostobjecttoscriptwithorigins)

---

> This article mainly covers mainly Win32/C++, and also covers some aspects of .NET/C# within frames.  For WinRT, see [Call native-side WinRT code from web-side code](./winrt-from-js.md).

Author: removed one

#### Why use `AddHostObjectToScript`?

  * When developing a WebView2 app, you may encounter a native object whose methods or properties you find useful. You might want to trigger these native object methods from web-side code, as a result of user interaction on the web side of your app. In addition, you might not want to re-implement your native objects' methods in your web-side code.  The `AddHostObjectToScript` API enables re-use of native-side code by web-side code. 

  * For example, there might be a native webcam API, which would require re-writing a large amount of code on the web side. Having the ability to call the native object's methods is quicker and more efficient than re-coding the object's methods on the web side of your app. In this case, your native-side code can pass the object to your app's web-side, JavaScript code, so that your JavaScript code can reuse the native API's methods.

Scenarios that may benefit from using host objects in script:

  * There is a keyboard API, and you want to call the `keyboardObject.showKeyboard` function from the web side.

>   * JavaScript is sandboxed, limiting its ability on the native side. For example, if you need to access a file on the native side, you must use the native file system. If you have a native object that's exposed to JavaScript via `AddHostObjectToScript`, you can use the object to manipulate files on the native file system.

Author: rewrote the scenario, eg:
Accessing the file system, not just the webpage sandbox, via JavaScript.  JavaScript is sandboxed, which prevents it from directly accessing the file system.  By using `AddHostObjectToScript` to create a native object that's exposed to JavaScript, you can use the host object to manipulate files on the file system, not just in the webpage sandbox.

This article uses the [Win32 sample app](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/WebView2APISample) to demonstrate some practical applications of `AddHostObjectToScript`.

<!-- ====================================================================== -->
## Step 1: Install Visual Studio, install git, clone the WebView2Samples repo, and open the solution

1.  Download and install [Microsoft Visual Studio](https://visualstudio.microsoft.com/) 2019 (version 16.11.10) or later, and other prerequisites as described in [Win32 sample app](../samples/webview2apissample.md).  The Win32 sample app was created using Visual Studio 2019, so to follow the example steps in this article, we recommend starting with Visual Studio 2019 rather than Visual Studio 2022.

1.  Clone the [WebView2Samples](https://github.com/MicrosoftEdge/WebView2Samples) repo.  The repo includes the Win32-specific WebView2 sample app.  For instructions, in a new window or tab, see [Win32 sample app](../samples/webview2apissample.md).

1.  Open Microsoft Visual Studio.  We recommend initially opening the Win32 sample by using Visual Studio 2019.

> 1.  In your local copy of the cloned `WebView2Samples` repo, open `WebView2Samples > SampleApps > WebView2Samples.sln`.  [WebView2Samples.sln](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2Samples.sln).) includes the `WebView2APISample` project, which is the Win32 sample app.  Keep the sample app solution open, to follow along with the rest of this article.

Author: restructured eg:
```
In your local copy of the cloned `WebView2Samples` repo, open `WebView2Samples` > `SampleApps` > [WebView2Samples.sln](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2Samples.sln).  `WebView2Samples.sln` includes the `WebView2APISample` project, which is the Win32 sample app.
```

<!-- ====================================================================== -->
## Step 2: Define the host object's COM interface using IDL

Define the host object's COM interface in an `.idl` file, like [HostObjectSample.idl](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2APISample/HostObjectSample.idl), to describe the methods and properties on the host object.  Here's how the Win32 sample app defines a host object with example methods and properties.

> To use the `AddHostObjectToScript` API, you first need to define a host object.  For Win32/C++, the host object definition in an `idl` file describes the exposed or wrapped properties and methods.

Author: @vbryh-msft - fyi if you were to use the [+-] Add A Suggestion button here, tip: 

In the + column between line #s and content, select the 3 lines of markdown by click&drag down.

> To use the `AddHostObjectToScript` API, you first need to define a host object.  For Win32/C++, the host object definition in an `idl` file describes the exposed or wrapped properties and methods.

Author: condensed eg:

First, use interface definition language (IDL) to define the host object's COM interface.  This host object definition in an `idl` file describes the exposed (or "wrapped") native-side properties and methods.

In this step, the Win32 sample app defines the host object's COM interface using interface definition language (IDL). This is demonstrated in [HostObjectSample.idl](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2APISample/HostObjectSample.idl).  The IDL (`.idl`) file *defines* an interface, but doesn't implement it.

> In the Win32 sample app project, the file [HostObjectSample.idl](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2APISample/HostObjectSample.idl) defines a COM object.  This step serves as a model showing how to define your own object in an IDL file.

Author: Condensed section

        m_propertyValue = stringValue;
        return S_OK;
    }
    ...

    STDMETHODIMP HostObjectSample::get_DateProperty(DATE* dateResult)
    {
        *dateResult = m_date;
        return S_OK;
    }
    
    STDMETHODIMP HostObjectSample::put_DateProperty(DATE dateValue)
    {
        m_date = dateValue;
        SYSTEMTIME systemTime;
        if (VariantTimeToSystemTime(dateValue, &systemTime))
    ...
    }
    
    STDMETHODIMP HostObjectSample::CreateNativeDate()
    {
        SYSTEMTIME systemTime;
        GetSystemTime(&systemTime);
        DATE date;
        if (SystemTimeToVariantTime(&systemTime, &date))
        {
            return put_DateProperty(date);
        }
        return E_UNEXPECTED;
    }
    ```

> 1.  Examine the date property, which we trace throughout this article.

Author: Examine `DateProperty`, which we trace throughout this article.

    
    ```cpp
            if (AreFileUrisEqual(sampleUri, uriTarget))
            {
                VARIANT remoteObjectAsVariant = {};
                m_hostObject.query_to<IDispatch>(&remoteObjectAsVariant.pdispVal);
                remoteObjectAsVariant.vt = VT_DISPATCH;
    ```

    Now that we have a variant of the object that is C++ code-friendly, the sample app's native-side code is ready to pass the host object to the app's web-side code.

1.  In the bottom line above, the `NavigationStarting` event handler then sets the remote object's variant type as `IDispatch`.

    ```cpp
                // We can call AddHostObjectToScript multiple times in a row without
                // calling RemoveHostObject first. This will replace the previous object
                // with the new object. In our case this is the same object and everything
                // is fine.
                CHECK_FAILURE(
                    m_webView->AddHostObjectToScript(L"sample", &remoteObjectAsVariant));
                remoteObjectAsVariant.pdispVal->Release();
            }
    ```

> 1.  The `NavigationStarting` event handler then passes the `VARIANT` to `AddHostObjectToScript` (above), names it `sample`, and enables the remote object as `VARIANT` (`&remoteObjectAsVariant`).

Author: unclear comment, recommend using [+-] Add Suggestion to provide a draft of a revision.

> 1.  The `NavigationStarting` event handler then passes the `VARIANT` to `AddHostObjectToScript` (above), names it `sample`, and enables the remote object as `VARIANT` (`&remoteObjectAsVariant`).

Reviewer: `VARIANT` named `sample` is passed in `NavigationStarting` event handler... I guess I did not like that event handler passes the `VARIANT`. But it seems the same in line 289 - that handler doing the job. Not sure if it makes sense why I'm confused. You can leave it as is if it is pattern which we use or rewrite if I was able to explain my POV

> 1.  The `NavigationStarting` event handler then passes the `VARIANT` to `AddHostObjectToScript` (above), names it `sample`, and enables the remote object as `VARIANT` (`&remoteObjectAsVariant`).

Reviewer: ```suggestion
1.  In `NavigationStarting` event handler the `VARIANT` to `AddHostObjectToScript` (above) is passed with the name `sample`.
```

> 1.  The `NavigationStarting` event handler then passes the `VARIANT` to `AddHostObjectToScript` (above), names it `sample`, and enables the remote object as `VARIANT` (`&remoteObjectAsVariant`).

Author: Changed to:
1. Above, in the `NavigationStarting` event handler, the `VARIANT` is passed to `AddHostObjectToScript`, using the name `sample`.

<!-- ----------------------------------- -->

##### [.NET in a frame](#tab/dotnetframe)

> To add the host object to an iframe, we'll use an overload of `CoreWebView2Frame.AddHostObjectToScript` which takes an `origins` parameter.

Author: To add the host object to an iframe, we'll use `CoreWebView2Frame.AddHostObjectToScript`, which takes an `origins` parameter.

#### Using the sample app

You can experiment with using and modifying the Win32 sample app.  Then follow the same pattern in your own app:

1.  Create a host object in your app's native-side code.
1.  Pass the host object to your app's web-side code.
1.  Use the host object from the app's web-side code.

To find out what other APIs there are in the host object ecosystem, see [WebView2 Win32 C++ ICoreWebView2](/microsoft-edge/webview2/reference/win32/icorewebview2#addhostobjecttoscript).

<!-- ====================================================================== -->
## API Reference overview

See [Host/web object sharing](../concepts/overview-features-apis.md#hostweb-object-sharing) in _Overview of WebView2 features and APIs_.

<!-- ====================================================================== -->
## See also

* [Web/native interop](../concepts/overview-features-apis.md#webnative-interop) in _Overview of WebView2 features and APIs_.
* [Using frames in WebView2 apps](../concepts/frames.md)

> * [Call native-side WinRT code from web-side code](./winrt-from-js.md)

Author: commit-specific view: does "Call native-side WinRT code from web-side code" link work?  
https://github.com/MicrosoftDocs/edge-developer/blob/49df45ef4d1bcf01447550ff5d21386ce1f47333/microsoft-edge/webview2/how-to/hostobject.md#see-also

I believe it is correct & working.  Try Refresh in rendered page for hostobject via View link on main PR page.

also try Find 2x on hostobject page:   Call native-side WinRT code from web-side code - try clicking both links.

> * [Call native-side WinRT code from web-side code](./winrt-from-js.md)

Author: Jason clicked the two instances of this link in the rendered page and they work.
